idp/idpclient/client.go

```diff
@@ -0,0 +1,171 @@
+These external users have been successfully authenticated by an external identity provider like google, facebook,...
+(if the admin configured them as external identity provider) but have not been added explicitly to the list of known users of the tenant.
+So you don't know much about external users except that they provided valid credentials to one of the configured external
+identity providers. For example if google is configured as an external identity provider anyone with a google account
+can sign in successfully.
+In most of the cases you don't want these users to access your business functions. Exceptions are business cases
+in which external users participate such as sharing information with identified but otherwise external users.
+Inspect the scim.Principal after a successful validation to distinguish external from internal users
+(cf. documentation of scim.Principal for further information).
+*/
+func (c *client) Validate(ctx context.Context, systemBaseUri string, tenantId string, authSessionId string) (*scim.Principal, error) {
+	cacheKey := fmt.Sprintf("%s/%s", tenantId, authSessionId)
+	co, found := c.principalCache.Get(cacheKey)
+	if found {
+		p := co.(scim.Principal)
+		return &p, nil
+	}
+
+	baseUri, baseParseErr := url.Parse(systemBaseUri)
+	if baseParseErr != nil {
+		return nil, baseParseErr
+	}
+	validatePath, _ := url.Parse("/identityprovider/validate?allowExternalValidation=true")
+	validateEndpoint := baseUri.ResolveReference(validatePath)
+
+	req, nRErr := http.NewRequestWithContext(ctx, http.MethodGet, validateEndpoint.String(), nil)
+	if nRErr != nil {
+		return nil, fmt.Errorf("can't create http request for '%s' because: %v", validateEndpoint, nRErr)
+	}
+	req.Header.Set("Authorization", "Bearer "+authSessionId)
+
+	resp, doErr := c.httpClient.Do(req)
+	if doErr != nil {
+		return nil, fmt.Errorf("error calling http GET on '%s' because: %w", validateEndpoint, doErr)
+	}
+	defer resp.Body.Close()
+
+	switch resp.StatusCode {
+	case http.StatusOK:
+		var p scim.Principal
+		if err := json.NewDecoder(resp.Body).Decode(&p); err != nil {
+			return nil, fmt.Errorf("response from Identityprovider '%s' is no valid JSON because: %v", validateEndpoint, err)
+		}
+		var validFor time.Duration = 0
+		cacheControlHeader := resp.Header.Get("Cache-Control")
+		matches := maxAgeRegex.FindStringSubmatch(cacheControlHeader)
+		if matches != nil {
+			d, err := time.ParseDuration(matches[1] + "s")
+			if err == nil {
+				validFor = d
+			}
+		}
+		if validFor > 0 {
+			c.principalCache.Set(cacheKey, p, validFor)
+		}
+		return &p, nil
+	case http.StatusUnauthorized:
+		_, _ = ioutil.ReadAll(resp.Body) // client must read to EOF and close body cf. https://godoc.org/net/http#Client
+		return nil, nil
+	default:
+		responseMsg, _ := ioutil.ReadAll(resp.Body)
+		return nil, fmt.Errorf(fmt.Sprintf("Identityprovider '%s' returned HTTP-Statuscode '%d' and message '%s'",
```

Comment:
fmt.Sprintf is not needed, as you already use fmt.Errorf.

Reply:
Done